### PR TITLE
LPS-61319 Users in a User Group do not inherit role permissions

### PR DESCRIPTION
**Pull request by pei-jung**
/cc @drewbrokke @jorgeferrer

---
This change is [Reviewable](https://reviewable.io/reviews/topolik/liferay-portal/355)

### CODE REVIEW

File: portal-impl/src/com/liferay/portal/security/permission/UserBagFactoryImpl.java [hunk 1]
```diff
@@ -46,9 +49,6 @@ public UserBag create(long userId) throws PortalException {
 		}
 
 		try {
-			List<Group> userGroups = GroupLocalServiceUtil.getUserGroups(
-				userId, true);
-
```
Comment (topolik) [inline portal-impl/src/com/liferay/portal/security/permission/UserBagFactoryImpl.java]: This line above should return the `Group` from `UserGroup` where user is member of. **But it's not!** Here we must search for the root cause!

Then there should be no need for all the code below.

Please notice the last `UNION` in the generated SQL:
```sql
(SELECT DISTINCT
   Group_.groupId                                AS groupId,
   REPLACE(Group_.name, ' LFR_ORGANIZATION', '') AS groupName,
   Group_.type_                                  AS groupType,
   Group_.friendlyURL                            AS groupFriendlyURL
 FROM Group_
   INNER JOIN Users_Groups ON (Users_Groups.groupId = Group_.groupId)
 WHERE (Group_.liveGroupId = 0) AND (Users_Groups.userId = ?) AND
       ((Group_.classNameId = ?) OR (Group_.classNameId = ?)) AND
       (Group_.companyId = ?) AND (Group_.parentGroupId != ?) AND
       (Group_.liveGroupId = 0) AND (Group_.groupKey != 'Control Panel') AND
       (lower(Group_.name) LIKE ? OR ? IS NULL) AND
       (lower(Group_.description) LIKE ? OR ? IS NULL) AND (Group_.type_ != 4))
UNION (SELECT DISTINCT
         Group_.groupId                                AS groupId,
         REPLACE(Group_.name, ' LFR_ORGANIZATION', '') AS groupName,
         Group_.type_                                  AS groupType,
         Group_.friendlyURL                            AS groupFriendlyURL
       FROM Group_
         INNER JOIN Users_Orgs ON (Users_Orgs.organizationId = Group_.classPK)
       WHERE (Users_Orgs.userId = ?) AND (Group_.classNameId = ?) AND
             (Group_.companyId = ?) AND (Group_.parentGroupId != ?) AND
             (Group_.liveGroupId = 0) AND (Group_.groupKey != 'Control Panel')
             AND (lower(Group_.name) LIKE ? OR ? IS NULL) AND
             (lower(Group_.description) LIKE ? OR ? IS NULL) AND
             (Group_.type_ != 4))
UNION (SELECT DISTINCT
         Group_.groupId                                AS groupId,
         REPLACE(Group_.name, ' LFR_ORGANIZATION', '') AS groupName,
         Group_.type_                                  AS groupType,
         Group_.friendlyURL                            AS groupFriendlyURL
       FROM Group_
         INNER JOIN Groups_Orgs ON (Groups_Orgs.groupId = Group_.groupId)
         INNER JOIN Users_Orgs
           ON (Users_Orgs.organizationId = Groups_Orgs.organizationId)
       WHERE (Group_.liveGroupId = 0) AND (Users_Orgs.userId = ?) AND
             (Group_.classNameId = ?) AND (Group_.companyId = ?) AND
             (Group_.parentGroupId != ?) AND (Group_.liveGroupId = 0) AND
             (Group_.groupKey != 'Control Panel') AND
             (lower(Group_.name) LIKE ? OR ? IS NULL) AND
             (lower(Group_.description) LIKE ? OR ? IS NULL) AND
             (Group_.type_ != 4))
UNION (SELECT DISTINCT
         Group_.groupId                                AS groupId,
         REPLACE(Group_.name, ' LFR_ORGANIZATION', '') AS groupName,
         Group_.type_                                  AS groupType,
         Group_.friendlyURL                            AS groupFriendlyURL
       FROM Group_
         INNER JOIN Groups_UserGroups
           ON (Groups_UserGroups.groupId = Group_.groupId)
         INNER JOIN Users_UserGroups
           ON (Users_UserGroups.userGroupId = Groups_UserGroups.userGroupId)
       WHERE (Group_.liveGroupId = 0) AND (Users_UserGroups.userId = ?) AND
             (Group_.classNameId = ?) AND (Group_.companyId = ?) AND
             (Group_.parentGroupId != ?) AND (Group_.liveGroupId = 0) AND
             (Group_.groupKey != 'Control Panel') AND
             (lower(Group_.name) LIKE ? OR ? IS NULL) AND
             (lower(Group_.description) LIKE ? OR ? IS NULL) AND
             (Group_.type_ != 4))
ORDER BY groupName ASC
```

**Comment by topolik** [comment]
Please try to find out why `List<Group> userGroups = GroupLocalServiceUtil.getUserGroups(userId, true);` doesn't return also UserGroup groups and resend. (My guess is `parentGroupId`, `classNameId` or `Group_.type_`)

Thank you.

File: portal-impl/src/com/liferay/portal/security/permission/UserBagFactoryImpl.java (hunk 1 above)
Comment (pei-jung) [inline portal-impl/src/com/liferay/portal/security/permission/UserBagFactoryImpl.java]: In the last union, the query only gets the sites that are associated with the user groups that the user is a member of, but not the user groups. Added a new query to join usergroup groups and sent you a new PR.

**Comment by pei-jung** [comment]
PR resent as https://github.com/topolik/liferay-portal/pull/361. Thanks!

**Pull request closed without merging**